File: FFRv2/app/src/main/java/com/devingotaswitch/rankings/domain/Draft.java
package com.devingotaswitch.rankings.domain;

import android.app.Activity;
import android.content.Context;
import android.support.design.widget.Snackbar;
import android.view.View;

import com.andrognito.flashbar.Flashbar;
import com.devingotaswitch.appsync.AppSyncHelper;
import com.devingotaswitch.fileio.LocalSettingsHelper;
import com.devingotaswitch.utils.Constants;
import com.devingotaswitch.utils.GeneralUtils;

import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Draft {

    private final List<String> draftedPlayers;
    private final Map<String, Integer> myPlayers;
    private final List<Player> myQbs;
    private final List<Player> myRbs;
    private final List<Player> myWrs;
    private final List<Player> myTes;
    private final List<Player> myDsts;
    private final List<Player> myKs;

    private double draftValue;

    public Draft() {
        draftedPlayers = new ArrayList<>();
        myQbs = new ArrayList<>();
        myRbs = new ArrayList<>();
        myWrs = new ArrayList<>();
        myTes = new ArrayList<>();
        myDsts = new ArrayList<>();
        myKs = new ArrayList<>();
        myPlayers = new HashMap<>();
        draftValue = 0.0;
    }

    public List<String> getDraftedPlayers() {
        return draftedPlayers;
    }

    public Map<String, Integer> getMyPlayers() {
        return myPlayers;
    }

    public List<Player> getMyQbs() {
        return myQbs;
    }

    public List<Player> getMyRbs() {
        return myRbs;
    }

    public List<Player> getMyWrs() {
        return myWrs;
    }

    public List<Player> getMyTes() {
        return myTes;
    }

    public List<Player> getMyDsts() {
        return myDsts;
    }

    public List<Player> getMyKs() {
        return myKs;
    }

    public List<Player> getPlayersDraftedForPos(String position) {
        switch (position) {
            case Constants.QB:
                return getMyQbs();
            case Constants.RB:
                return getMyRbs();
            case Constants.WR:
                return getMyWrs();
            case Constants.TE:
                return getMyTes();
            case Constants.DST:
                return getMyDsts();
            case Constants.K:
                return getMyKs();
        }
        return new ArrayList<>();
    }

    public List<Player> getPlayersWithSameByeAndPos(Player player, Rankings rankings) {
        return getPlayersWithSameBye(getPlayersDraftedForPos(player.getPosition()),
                rankings, rankings.getTeam(player).getBye());
    }

    public List<Player> getPlayersWithSameBye(Player player, Rankings rankings) {
        List<Player> allMyPicks = new ArrayList<>();
        allMyPicks.addAll(getMyQbs());
        allMyPicks.addAll(getMyRbs());
        allMyPicks.addAll(getMyWrs());
        allMyPicks.addAll(getMyTes());
        allMyPicks.addAll(getMyDsts());
        allMyPicks.addAll(getMyKs());
        return getPlayersWithSameBye(allMyPicks, rankings, rankings.getTeam(player).getBye());
    }

    private List<Player> getPlayersWithSameBye(List<Player> toCheck, Rankings rankings, String bye) {
        List<Player> sameBye = new ArrayList<>();
        for (Player posPlayer : toCheck) {
            if (bye.equals(rankings.getTeam(posPlayer).getBye())) {
                sameBye.add(posPlayer);
            }
        }
        return sameBye;
    }

    public double getDraftValue() {
        return draftValue;
    }

    public double getTotalPAA() {
        return getQBPAA() + getRBPAA() + getWRPAA() + getTEPAA() + getDSTPAA() + getKPAA();
    }

    public double getTotalXVal() {
        return getQBXval() + getRBXval() + getWRXval() + getTEXval() + getDSTXval() + getKXval();
    }

    public double getTotalVoLS() {
        return getQBVoLS() + getRBVoLS() + getWRVoLS() + getTEVoLS() + getDSTVoLS() + getKVoLS();
    }

    public double getQBXval() {
        return getXValForPos(myQbs);
    }

    public double getWRXval() {
        return getXValForPos(myWrs);
    }

    public double getRBXval() {
        return getXValForPos(myRbs);
    }

    public double getTEXval() {
        return getXValForPos(myTes);
    }

    public double getDSTXval() {
        return getXValForPos(myDsts);
    }

    public double getKXval() {
        return getXValForPos(myKs);
    }

    private double getXValForPos(List<Player> players) {
        double posXval = 0.0;
        for (Player player : players) {
            posXval += player.getxVal();
        }
        return posXval;
    }

    public double getQBVoLS() {
        return getVoLSForPos(myQbs);
    }

    public double getRBVoLS() {
        return getVoLSForPos(myRbs);
    }

    public double getWRVoLS() {
        return getVoLSForPos(myWrs);
    }

    public double getTEVoLS() {
        return getVoLSForPos(myTes);
    }

    public double getDSTVoLS() {
        return getVoLSForPos(myDsts);
    }

    public double getKVoLS() {
        return getVoLSForPos(myKs);
    }

    private double getVoLSForPos(List<Player> players) {
        double posVoLS = 0.0;
        for (Player player : players) {
            posVoLS += player.getVOLS();
        }
        return posVoLS;
    }

    public double getQBPAA() {
        return getPAAForPos(myQbs);
    }

    public double getWRPAA() {
        return getPAAForPos(myWrs);
    }

    public double getRBPAA() {
        return getPAAForPos(myRbs);
    }

    public double getTEPAA() {
        return getPAAForPos(myTes);
    }

    public double getDSTPAA() {
        return getPAAForPos(myDsts);
    }

    public double getKPAA() {
        return getPAAForPos(myKs);
    }

    private double getPAAForPos(List<Player> players) {
        double posPAA = 0.0;
        for (Player player : players) {
            posPAA += player.getPaa();
        }
        return posPAA;
    }

    public void draftPlayer(Player player, int teamCount, int auctionBudget, boolean myPick, int cost) {
        draftedPlayers.add(player.getUniqueId());
        if (myPick) {
            this.myPlayers.put(player.getUniqueId(), cost);
            this.draftValue += (player.getAuctionValueCustom(teamCount, auctionBudget) - (double) cost);
            switch (player.getPosition()) {
                case Constants.QB:
                    myQbs.add(player);
                    break;
                case Constants.RB:
                    myRbs.add(player);
                    break;
                case Constants.WR:
                    myWrs.add(player);
                    break;
                case Constants.TE:
                    myTes.add(player);
                    break;
                case Constants.DST:
                    myDsts.add(player);
                    break;
                case Constants.K:
                    myKs.add(player);
                    break;
            }
        }
    }

    private void unDraftPlayer(Player player, Rankings rankings) {
        draftedPlayers.remove(player.getUniqueId());
        if (isDraftedByMe(player)) {
            int cost = myPlayers.get(player.getUniqueId());
            this.myPlayers.remove(player.getUniqueId());
            this.draftValue -= (player.getAuctionValueCustom(rankings) - (double) cost);
            switch (player.getPosition()) {
                case Constants.QB:
                    myQbs.remove(player);
                    break;
                case Constants.RB:
                    myRbs.remove(player);
                    break;
                case Constants.WR:
                    myWrs.remove(player);
                    break;
                case Constants.TE:
                    myTes.remove(player);
                    break;
                case Constants.DST:
                    myDsts.remove(player);
                    break;
                case Constants.K:
                    myKs.remove(player);
                    break;
            }
        }
    }

    public boolean isDraftedByMe(Player player) {
        return myPlayers.containsKey(player.getUniqueId());
    }

    public boolean isDrafted(Player player) { return draftedPlayers.contains(player.getUniqueId()); }

    public String draftedToSerializedString() {
        StringBuilder draftedStr = new StringBuilder();
        for (String key : draftedPlayers) {
            draftedStr.append(key)
                    .append(Constants.HASH_DELIMITER);
        }
        return draftedStr.toString();
    }

    public String myTeamToSerializedString() {
        StringBuilder myTeamStr = new StringBuilder();
        for (String key : myPlayers.keySet()) {
            int cost = myPlayers.get(key);
            myTeamStr.append(key)
                    .append(Constants.HASH_DELIMITER)
                    .append(cost)
                    .append(Constants.HASH_DELIMITER);
        }
        return myTeamStr.toString();
    }

    public List<Player> getSortedAvailablePlayersForPosition(String pos, Rankings rankings) {
        Comparator<Player> comparator = new Comparator<Player>() {
            @Override
            public int compare(Player a, Player b) {
                if (a.getProjection() > b.getProjection()) {
                    return -1;
                }
                if (a.getProjection() < b.getProjection()) {
                    return 1;
                }
                return 0;
            }
        };
        List<Player> players = new ArrayList<>();
        for (Player player : rankings.getPlayers().values()) {
            if (!isDrafted(player) && player.getPosition().equals(pos)) {
                players.add(player);
            }
        }
        Collections.sort(players, comparator);
        return players;
    }

    public double getPAANAvailablePlayersBack(List<Player> players, int limit) {
        int counter = 0;
        double paaLeft = 0.0;
        for (Player player : players) {
            paaLeft += player.getPaa();
            counter++;
            if (counter == limit) {
                break;
            }
        }
        return paaLeft;
    }

    public String getPAALeft(String pos, Rankings rankings) {
        DecimalFormat df = new DecimalFormat("#.#");
        String result = pos + "s: ";
        List<Player> players = getSortedAvailablePlayersForPosition(pos, rankings);

        result += df.format(getPAANAvailablePlayersBack(players, 3)) + "/";
        result += df.format(getPAANAvailablePlayersBack(players, 5)) + "/";
        result += df.format(getPAANAvailablePlayersBack(players, 10));

        if (result.endsWith("/")) {
            result = result.substring(0, result.length() - 1);
        }
        return result;
    }

    public void resetDraft(Context context, String leagueName) {
        myQbs.clear();
        myRbs.clear();
        myWrs.clear();
        myTes.clear();
        myDsts.clear();
        myKs.clear();
        myPlayers.clear();
        draftedPlayers.clear();
        draftValue = 0.0;
        LocalSettingsHelper.clearDraft(context, leagueName);
    }

    public void draftBySomeone(Rankings rankings, Player player, Activity act, View view, View.OnClickListener listener) {
        draftPlayer(player, rankings.getLeagueSettings().getTeamCount(), rankings.getLeagueSettings().getAuctionBudget(),false, 0);
        if (listener == null) {
            GeneralUtils.generateTextOnlyFlashbar(act, "Success!", player.getName() + " drafted by you", Flashbar.Gravity.BOTTOM)
                    .show();
        } else {
            Snackbar.make(view, player.getName() + " drafted", Snackbar.LENGTH_LONG).setAction("Undo", listener).show();
        }
        saveDraft(rankings, act);
        AppSyncHelper.incrementPlayerDraftCount(act, player.getUniqueId());
    }

    public void draftByMe(Rankings rankings, Player player, Activity act, int cost, View view, View.OnClickListener listener) {
        draftPlayer(player, rankings.getLeagueSettings().getTeamCount(), rankings.getLeagueSettings().getAuctionBudget(),true, cost);
        if (listener == null) {
            GeneralUtils.generateTextOnlyFlashbar(act, "Success!", player.getName() + " drafted by you", Flashbar.Gravity.BOTTOM)
                    .show();
        } else {
            Snackbar.make(view, player.getName() + " drafted by you", Snackbar.LENGTH_LONG).setAction("Undo", listener).show();
        }
        saveDraft(rankings, act);
        AppSyncHelper.incrementPlayerDraftCount(act, player.getUniqueId());
    }

    public void undraft(Rankings rankings, Player player, Activity act, View view) {
        unDraftPlayer(player, rankings);
        GeneralUtils.generateTextOnlyFlashbar(act, "Success!", player.getName() + " undrafted", Flashbar.Gravity.BOTTOM)
                .show();
        saveDraft(rankings, act);
        AppSyncHelper.decrementPlayerDraftCount(act, player.getUniqueId());
    }

    private void saveDraft(Rankings rankings, Activity act) {
        LocalSettingsHelper.saveDraft(act, rankings.getLeagueSettings().getName(), rankings.getDraft());
    }
}

File: FFRv2/app/src/main/java/com/devingotaswitch/rankings/domain/Draft.kt
package com.devingotaswitch.rankings.domain

import android.app.Activity
import android.content.Context
import com.andrognito.flashbar.Flashbar
import com.andrognito.flashbar.Flashbar.OnActionTapListener
import com.devingotaswitch.appsync.AppSyncHelper.decrementPlayerDraftCount
import com.devingotaswitch.appsync.AppSyncHelper.incrementPlayerDraftCount
import com.devingotaswitch.fileio.LocalSettingsHelper
import com.devingotaswitch.utils.Constants
import com.devingotaswitch.utils.FlashbarFactory.generateFlashbarWithUndo
import com.devingotaswitch.utils.FlashbarFactory.generateTextOnlyFlashbar
import java.text.DecimalFormat
import java.util.*

class Draft {
    val draftedPlayers: MutableList<String>
    val myPlayers: MutableMap<String, Int>
    val myQbs: MutableList<Player>
    val myRbs: MutableList<Player>
    val myWrs: MutableList<Player>
    val myTes: MutableList<Player>
    val myDsts: MutableList<Player>
    val myKs: MutableList<Player>
    var draftValue: Double
        private set

    fun getPlayersDraftedForPos(position: String?): List<Player> {
        when (position) {
            Constants.QB -> return myQbs
            Constants.RB -> return myRbs
            Constants.WR -> return myWrs
            Constants.TE -> return myTes
            Constants.DST -> return myDsts
            Constants.K -> return myKs
        }
        return ArrayList()
    }

    fun getPlayersWithSameByeAndPos(player: Player, rankings: Rankings): List<Player> {
        return getPlayersWithSameBye(getPlayersDraftedForPos(player.position),
                rankings, rankings.getTeam(player)!!.bye)
    }

    fun getPlayersWithSameBye(player: Player, rankings: Rankings): List<Player> {
        val allMyPicks: MutableList<Player> = ArrayList()
        allMyPicks.addAll(myQbs)
        allMyPicks.addAll(myRbs)
        allMyPicks.addAll(myWrs)
        allMyPicks.addAll(myTes)
        allMyPicks.addAll(myDsts)
        allMyPicks.addAll(myKs)
        return getPlayersWithSameBye(allMyPicks, rankings, rankings.getTeam(player)!!.bye)
    }

    private fun getPlayersWithSameBye(toCheck: List<Player>, rankings: Rankings, bye: String?): List<Player> {
        val sameBye: MutableList<Player> = ArrayList()
        for (posPlayer in toCheck) {
            if (bye == rankings.getTeam(posPlayer)?.bye) {
                sameBye.add(posPlayer)
            }
        }
        return sameBye
    }

    val totalPAA: Double
        get() = qBPAA + rBPAA + wRPAA + tEPAA + dstpaa + kpaa
    val totalXVal: Double
        get() = qBXval + rBXval + wRXval + tEXval + dstXval + kXval
    val totalVoLS: Double
        get() = qBVoLS + rBVoLS + wRVoLS + tEVoLS + dstVoLS + kVoLS
    val qBXval: Double
        get() = getXValForPos(myQbs)
    val wRXval: Double
        get() = getXValForPos(myWrs)
    val rBXval: Double
        get() = getXValForPos(myRbs)
    val tEXval: Double
        get() = getXValForPos(myTes)
    val dstXval: Double
        get() = getXValForPos(myDsts)
    val kXval: Double
        get() = getXValForPos(myKs)

    private fun getXValForPos(players: List<Player>): Double {
        var posXval = 0.0
        for (player in players) {
            posXval += player.xval
        }
        return posXval
    }

    val qBVoLS: Double
        get() = getVoLSForPos(myQbs)
    val rBVoLS: Double
        get() = getVoLSForPos(myRbs)
    val wRVoLS: Double
        get() = getVoLSForPos(myWrs)
    val tEVoLS: Double
        get() = getVoLSForPos(myTes)
    val dstVoLS: Double
        get() = getVoLSForPos(myDsts)
    val kVoLS: Double
        get() = getVoLSForPos(myKs)

    private fun getVoLSForPos(players: List<Player>): Double {
        var posVoLS = 0.0
        for (player in players) {
            posVoLS += player.vols
        }
        return posVoLS
    }

    val qBPAA: Double
        get() = getPAAForPos(myQbs)
    val wRPAA: Double
        get() = getPAAForPos(myWrs)
    val rBPAA: Double
        get() = getPAAForPos(myRbs)
    val tEPAA: Double
        get() = getPAAForPos(myTes)
    val dstpaa: Double
        get() = getPAAForPos(myDsts)
    val kpaa: Double
        get() = getPAAForPos(myKs)

    private fun getPAAForPos(players: List<Player>): Double {
        var posPAA = 0.0
        for (player in players) {
            posPAA += player.paa
        }
        return posPAA
    }

    fun draftPlayer(player: Player, teamCount: Int, auctionBudget: Int, myPick: Boolean, cost: Int) {
        draftedPlayers.add(player.uniqueId)
        if (myPick) {
            myPlayers[player.uniqueId] = cost
            draftValue += player.getAuctionValueCustom(teamCount, auctionBudget) - cost.toDouble()
            when (player.position) {
                Constants.QB -> myQbs.add(player)
                Constants.RB -> myRbs.add(player)
                Constants.WR -> myWrs.add(player)
                Constants.TE -> myTes.add(player)
                Constants.DST -> myDsts.add(player)
                Constants.K -> myKs.add(player)
            }
        }
    }

    private fun unDraftPlayer(player: Player, rankings: Rankings) {
        draftedPlayers.remove(player.uniqueId)
        if (isDraftedByMe(player)) {
            val cost = myPlayers[player.uniqueId]!!
            myPlayers.remove(player.uniqueId)
            draftValue -= player.getAuctionValueCustom(rankings) - cost.toDouble()
            when (player.position) {
                Constants.QB -> myQbs.remove(player)
                Constants.RB -> myRbs.remove(player)
                Constants.WR -> myWrs.remove(player)
                Constants.TE -> myTes.remove(player)
                Constants.DST -> myDsts.remove(player)
                Constants.K -> myKs.remove(player)
            }
        }
    }

    fun isDraftedByMe(player: Player): Boolean {
        return myPlayers.containsKey(player.uniqueId)
    }

    fun isDrafted(player: Player): Boolean {
        return draftedPlayers.contains(player.uniqueId)
    }

    fun draftedToSerializedString(): String {
        val draftedStr = StringBuilder()
        for (key in draftedPlayers) {
            draftedStr.append(key)
                    .append(Constants.HASH_DELIMITER)
        }
        return draftedStr.toString()
    }

    fun myTeamToSerializedString(): String {
        val myTeamStr = StringBuilder()
        for (key in myPlayers.keys) {
            val cost: Int = myPlayers[key]!!
            myTeamStr.append(key)
                    .append(Constants.HASH_DELIMITER)
                    .append(cost)
                    .append(Constants.HASH_DELIMITER)
        }
        return myTeamStr.toString()
    }

    fun getSortedAvailablePlayersForPosition(pos: String, rankings: Rankings): List<Player> {
        val comparator = Comparator { a: Player, b: Player -> b.projection.compareTo(a.projection) }
        val players: MutableList<Player> = ArrayList()
        for (player in rankings.players.values) {
            if (!isDrafted(player) && player.position == pos) {
                players.add(player)
            }
        }
        Collections.sort(players, comparator)
        return players
    }

    fun getPAANAvailablePlayersBack(players: List<Player>, limit: Int): Double {
        var counter = 0
        var paaLeft = 0.0
        for (player in players) {
            paaLeft += player.paa
            counter++
            if (counter == limit) {
                break
            }
        }
        return paaLeft
    }

    fun getPAALeft(pos: String, rankings: Rankings): String {
        val df = DecimalFormat("#.#")
        var result = pos + "s: "
        val players = getSortedAvailablePlayersForPosition(pos, rankings)
        result += df.format(getPAANAvailablePlayersBack(players, 3)) + "/"
        result += df.format(getPAANAvailablePlayersBack(players, 5)) + "/"
        result += df.format(getPAANAvailablePlayersBack(players, 10))
        if (result.endsWith("/")) {
            result = result.substring(0, result.length - 1)
        }
        return result
    }

    fun resetDraft(context: Context, leagueName: String) {
        myQbs.clear()
        myRbs.clear()
        myWrs.clear()
        myTes.clear()
        myDsts.clear()
        myKs.clear()
        myPlayers.clear()
        draftedPlayers.clear()
        draftValue = 0.0
        LocalSettingsHelper.clearDraft(context, leagueName)
    }

    fun draftBySomeone(rankings: Rankings, player: Player, act: Activity, listener: OnActionTapListener?) {
        draftPlayer(player, rankings.getLeagueSettings().teamCount, rankings.getLeagueSettings().auctionBudget, false, 0)
        if (listener == null) {
            generateTextOnlyFlashbar(act, "Success!", player.name + " drafted", Flashbar.Gravity.BOTTOM)
                    .show()
        } else {
            generateFlashbarWithUndo(act, "Success!", player.name + " drafted", Flashbar.Gravity.BOTTOM, listener)
                    .show()
        }
        saveDraft(rankings, act)
        incrementPlayerDraftCount(act, player.uniqueId)
    }

    fun draftByMe(rankings: Rankings, player: Player, act: Activity, cost: Int, listener: OnActionTapListener?) {
        draftPlayer(player, rankings.getLeagueSettings().teamCount, rankings.getLeagueSettings().auctionBudget, true, cost)
        if (listener == null) {
            generateTextOnlyFlashbar(act, "Success!", player.name + " drafted by you", Flashbar.Gravity.BOTTOM)
                    .show()
        } else {
            generateFlashbarWithUndo(act, "Success!", player.name + " drafted by you", Flashbar.Gravity.BOTTOM, listener)
                    .show()
        }
        saveDraft(rankings, act)
        incrementPlayerDraftCount(act, player.uniqueId)
    }

    fun undraft(rankings: Rankings, player: Player, act: Activity) {
        unDraftPlayer(player, rankings)
        generateTextOnlyFlashbar(act, "Success!", player.name + " undrafted", Flashbar.Gravity.BOTTOM)
                .show()
        saveDraft(rankings, act)
        decrementPlayerDraftCount(act, player.uniqueId)
    }

    private fun saveDraft(rankings: Rankings, act: Activity) {
        LocalSettingsHelper.saveDraft(act, rankings.getLeagueSettings().name, rankings.draft)
    }

    init {
        draftedPlayers = ArrayList()
        myQbs = ArrayList()
        myRbs = ArrayList()
        myWrs = ArrayList()
        myTes = ArrayList()
        myDsts = ArrayList()
        myKs = ArrayList()
        myPlayers = HashMap()
        draftValue = 0.0
    }
}